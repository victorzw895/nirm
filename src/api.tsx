import { createClient } from '@supabase/supabase-js'

export const supabase = createClient(import.meta.env.VITE_SUPABASE_URL, import.meta.env.VITE_SUPABASE_KEY)

const fetchAnimes = async ({limit, offset, year} = {limit: 20, offset: 0, year: 2000}) => {
    const response = await fetch(`${import.meta.env.VITE_ANIME_API}?page[limit]=${limit}&page[offset]=${offset}&filter[season_year]=${year}&sort=createdAt`);
    return await response.json();
}

type ImageSizes = 'original' | 'tiny' | 'small' | 'medium' | 'large'
interface Images extends Record<ImageSizes, string> {
    meta: {
        dimensions: Record<ImageSizes, {
            width: number,
            height: number
        }>
    }
}

export interface Anime {
    id: number,
    attributes: {
        'createdAt': string,
        'updatedAt': string,
        'slug': string,
        'synopsis': string,
        'description': string,
        'coverImageTopOffset': number,
        'titles': Record<'en' | 'en_jp' | 'ja_jp', string>,
        'canonicalTitle': string,
        'abbreviatedTitles': string[],
        'averageRating': string,
        'ratingFrequencies': Record<string, string>,
        'userCount': number,
        'favoritesCount': number,
        'startDate': string,
        'endDate': string,
        'nextRelease': null,
        'popularityRank': number,
        'ratingRank': number,
        'ageRating': string,
        'ageRatingGuide': string,
        'subtype': string,
        'status': string,
        'tba': null,
        'posterImage': Images,
        'coverImage': Images,
        'episodeCount': number,
        'episodeLength': number,
        'totalLength': number,
        'youtubeVideoId': string,
        'showType': string,
        'nsfw': boolean
    },
    rank: number | null,
    stars: number,
    isWatched: boolean,
    seasonYear: number,
    watchlist: boolean,
}

const getSeasonYears = async () => {
    let { data }: { data: number[]} = await supabase
        .rpc('getseasonyears')

    return data;
}

export type AnimesByYear = Record<string, Anime[]>

const getAnimesByYear = async (year: number) => {
    const { data } = await supabase
        .from('AnimeList')
        .select()
        .eq('seasonYear', year)
        .order('id', { ascending: true })
    
    return data as Anime[];
}

const getAllAnimes = async () => {
    const dbSeasonYears = await getSeasonYears()
    const animes = await Promise.all(dbSeasonYears.map(async (year) => await getAnimesByYear(year)))

    const animesByYear: {[x: string]: Anime[]} = dbSeasonYears.reduce((acc, year, index) => ({
        ...acc,
        [year.toString()]: animes[index]
    }), {})

    return animesByYear;
}

const getAnimeList = async () => {
    let data: AnimesByYear;
    const dbAnimes = await getAllAnimes();

    if (dbAnimes && Object.keys(dbAnimes).length) {
        data = dbAnimes
    }
    else {
        const { data: fetchInitialAnimesData }: {data: Partial<Anime>[]} = await fetchAnimes()

        data = {
            '2000': fetchInitialAnimesData.map(anime => 
                ({
                    id: anime.id,
                    attributes: anime.attributes,
                    rank: null,
                    stars: 0,
                    isWatched: false,
                    seasonYear: 2000,
                }) as Anime
            )
        }
        
        insertAnime(data['2000'])
    };

    return data;
}

const getAnimeRankedList = async () => {
    const { data } = await supabase
        .from('AnimeList')
        .select()
        .eq('isWatched', true)
        .order('rank', { ascending: true });

    return data as Anime[];
}

const getAnimeWatchList = async () => {
    const { data } = await supabase
        .from('AnimeList')
        .select()
        .eq('watchlist', true)
        .order('rank', { ascending: true });

    return data as Anime[];
}

const insertAnime = async (anime: Anime[]) => {
    const { error } = await supabase
        .from('AnimeList')
        .insert(anime)
}

const upsertAnimeWatched = async (anime: Anime | Anime[]) => {
    const { data }: {data: Anime[]} = await supabase
        .from('AnimeList')
        .upsert(anime)
        .select()
    
    return data;
}

const updateAnimeWatched = async (anime: Anime) => {
    const { data }: {data: Anime} = await supabase
        .from('AnimeList')
        .update(anime)
        .eq('id', anime.id)
        .select()
        .single()
    
    return data;
}

export {
    fetchAnimes,
    getAnimeList,
    getAnimeRankedList,
    getSeasonYears,
    insertAnime,
    upsertAnimeWatched,
    updateAnimeWatched
}