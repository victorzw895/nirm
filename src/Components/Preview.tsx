import { Component, createSignal } from 'solid-js';
import moment from 'moment';
import { Anime } from '../api';

const [selectedAnime, setSelectedAnime] = createSignal<Anime | Partial<Anime> | null>(null);

interface PreviewProps {
  animeWatched: () => void,
  nextAnime: () => void,
}

const Preview: Component<PreviewProps> = (props) => {
  return (
    <section class='grid auto-rows-auto content-center col-span-2'>
      <div class='flex items-center mx-1'>
        <img class='max-h-80 m-2' src={selectedAnime()?.attributes.posterImage.small}
        />
        <div class='ml-2'>
          <p>{selectedAnime()?.attributes.titles.en_jp}</p>
          <p class='text-xs'>{selectedAnime()?.attributes.titles.en}</p>
          <p class='text-xs'>Rating {selectedAnime()?.attributes.ageRatingGuide}</p>
          <p class='text-xs'>Release Date: {moment(selectedAnime()?.attributes.startDate).format('YYYY')}</p>
        </div>
      </div>
      <div class='ml-2'>
        <p>Description</p>
        <p class='text-xs'>{
            selectedAnime()?.attributes.description
        }</p>
      </div>
      <div class='ml-2 flex justify-evenly'>
      <button
        class={`border rounded my-3 p-2 ${selectedAnime()?.isWatched ? 'bg-gray-400 border-gray-400' : 'border-neutral-800'}`}
        onClick={props.animeWatched}
      >
        Watched
      </button>
      <button
        class={`border rounded my-3 p-2 border-neutral-800`}
        onClick={props.nextAnime}
      >
        Next
      </button>
      </div>
    </section>
  )
}

export { selectedAnime, setSelectedAnime };

export default Preview;