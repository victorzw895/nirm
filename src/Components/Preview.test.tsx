import { fireEvent, render } from "@solidjs/testing-library";
import Preview, { setSelectedAnime, PreviewProps } from "./Preview";
import { Anime } from '../api';

const previewProps: PreviewProps = {
  animeWatched: () => {},
  nextAnime: () => {},
};

const selectedAnime = {
  id: 2,
  attributes: {
      description: 'description text',
      titles: {
        'en': 'en title',
        'en_jp': 'en jap title',
      },
      startDate: '2020',
      endDate: '2020',
      posterImage: {
        small: 'small-image-source-url',
      },
      ageRatingGuide: 'age rating guide',
  },
  rank: null,
  stars: 0,
  isWatched: false,
  seasonYear: 2011,
  watchlist: false,
} as Partial<Anime>

describe("<Preview />", () => {
  describe("Unwatched Anime Preview", () => {
    beforeEach(() => {
      setSelectedAnime(selectedAnime);
    })
  
    it("Displays Preview Text Content", () => {
      const { getByTestId, unmount } = render(() => <Preview {...previewProps} />);
      const textContents = getByTestId("text-content").querySelectorAll('p');
      expect(textContents[0]).toHaveTextContent(/en jap title/);
      expect(textContents[1]).toHaveTextContent(/en title/);
      expect(textContents[2]).toHaveTextContent(/age rating guide/);
      expect(textContents[3]).toHaveTextContent(/2020/);
      unmount();
    });
  
    it("Displays Preview Image Content", () => {
      const { getByTestId, unmount } = render(() => <Preview {...previewProps} />);
      const poster = getByTestId("poster")
      expect(poster).toHaveAttribute('src', 'small-image-source-url');
      unmount();
    });
  
    it("Displays Preview Description Content", () => {
      const { getByTestId, unmount } = render(() => <Preview {...previewProps} />);
      const description = getByTestId("description")
      expect(description).toHaveTextContent(selectedAnime.attributes.description);
      unmount();
    });
  
    it("Displays Preview Watched Button", () => {
      const { getByTestId, unmount } = render(() => <Preview {...previewProps} />);
      const watchedButton = getByTestId("watched-button");
      expect(watchedButton).toBeInTheDocument();
      expect(watchedButton).toHaveClass('bg-dark');
      unmount();
    });

    it("Displays Preview Next Button", () => {
      const { getByTestId, unmount } = render(() => <Preview {...previewProps} />);
      const nextButton = getByTestId("next-button");
      expect(nextButton).toBeInTheDocument();
      unmount();
    });
  })

  it("Displays Preview Watched Button - greyed out if watched", () => {
    setSelectedAnime({...selectedAnime, isWatched: true});
    const { getByTestId, unmount } = render(() => <Preview {...previewProps} />);
    const watchedButton = getByTestId("watched-button");
    expect(watchedButton).toBeInTheDocument();
    expect(watchedButton).toHaveClass('bg-light text-darkest');
    unmount();
  });

  it("Adds Preview Anime to watched list", () => {
    const animeWatchedMock = vi.fn();
    const { getByTestId, unmount } = render(() => <Preview {...{...previewProps, animeWatched: animeWatchedMock}} />);
    const watchedButton = getByTestId("watched-button");
    fireEvent.click(watchedButton);
    expect(animeWatchedMock).toHaveBeenCalledTimes(1);
    unmount();
  })

  it("Displays Next Preview Anime, when clicking next", () => {
    const nextAnimeMock = vi.fn();
    const { getByTestId, unmount } = render(() => <Preview {...{...previewProps, nextAnime: nextAnimeMock}} />);
    const nextButton = getByTestId("next-button");
    fireEvent.click(nextButton);
    expect(nextAnimeMock).toHaveBeenCalledTimes(1);
    unmount();
  })

  it("renders Preview", () => {
    const { container, unmount } = render(() => <Preview {...previewProps} />);
    expect(container).toMatchSnapshot();
    unmount();
  });
});