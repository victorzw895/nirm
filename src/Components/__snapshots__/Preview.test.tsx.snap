// Vitest Snapshot v1

exports[`<Preview /> > renders Preview 1`] = `
<div>
  <section
    class="col-span-2 self-center"
  >
    <div
      class="flex items-center mx-1"
    >
      <img
        class="max-h-80 m-2"
        data-testid="poster"
        src="small-image-source-url"
      />
      <div
        class="ml-2"
        data-testid="text-content"
      >
        <p>
          en jap title
        </p>
        <p
          class="text-xs"
        >
          en title
        </p>
        <p
          class="text-xs"
        >
          Rating 
          age rating guide
        </p>
        <p
          class="text-xs"
        >
          Release Date: 
          2020
        </p>
      </div>
    </div>
    <div
      class="ml-2 min-h-[10rem]"
    >
      <p>
        Description
      </p>
      <p
        class="text-xs line-clamp-8"
        data-testid="description"
      >
        description text
      </p>
    </div>
    <div
      class="ml-2 flex justify-evenly"
    >
      <button
        class="rounded my-3 p-2 bg-light text-darkest"
        data-testid="watched-button"
      >
        Watched
      </button>
      <button
        class="rounded my-3 p-2 bg-dark"
        data-testid="next-button"
      >
        Next
      </button>
    </div>
  </section>
</div>
`;
